api/app/signals/apps/signals/utils/map.py

```diff
@@ -0,0 +1,78 @@
+import io
+import urllib.request
+from math import asinh, pi, radians, tan
+
+from PIL import Image
+
+TILE_SIZE = 256
+
+
+#  https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Tile_numbers_to_lon..2Flat._3
+class MapGenerator:
+    def _deg2num(self, lat_deg, lon_deg, zoom):
+        lat_rad = radians(lat_deg)
+        n = 2.0 ** zoom
+        xtile = (lon_deg + 180.0) / 360.0 * n
+        ytile = (1.0 - asinh(tan(lat_rad)) / pi) / 2.0 * n
+        return (xtile, ytile)
+
+    def deg2num(self, lat_deg, lon_deg, zoom):
+        (xtile, ytile) = self._deg2num(lat_deg, lon_deg, zoom)
+        return (int(xtile), int(ytile))
+
+    def deg2num_pixel(self, lat_deg, lon_deg, zoom):
+        (xtile, ytile) = self._deg2num(lat_deg, lon_deg, zoom)
+        xpixel = int((xtile % 1) * TILE_SIZE)
+        ypixel = int((ytile % 1) * TILE_SIZE)
+        return (xpixel, ypixel)
+
+    def load_image(self, url_template, zoom, x, y):
+        img = Image.new("RGBA", (3*TILE_SIZE, 3*TILE_SIZE), 0)
+        try:
+            for i in range(-1, 2, 1):
+                for j in range(-1, 2, 1):
+                    url = url_template.format(zoom=zoom, x=x+i, y=y+j)
+                    with urllib.request.urlopen(url) as response:
+                        offset = ((i+1) * TILE_SIZE, (j+1) * TILE_SIZE)
+                        img.paste(
+                            Image.open(io.BytesIO(response.read())),
+                            offset
+                        )
+        except Exception:
+            pass  # use empty image in case of errors
+
+        return img
+
+    def _clip_coordinates(self, input):
+        rounded = round(input)
+        if rounded < 0:
+            return 0
+        if rounded > 3 * TILE_SIZE:
+            return 3 * TILE_SIZE
```

Comment:
This will probably have to change given my comment above --- you may have to separately handle the X and Y case (unless you change the magic number from 3 to 5).

Reply:
See previous comments, boundaries are correct.